Migrate ADAL to watchOS target and expose API for watchOS authentication

ertang2015 opened the pull request:
This pull request includes following changes:
- Created ADAL framework for WatchOS target and resolved all necessary compilation issues.
- Exposed necessary APIs for watchOS taking auth info from iOS and then authenticate on its own.

@SergeyADoroshenko @GaryCaldwell

ertang2015 committed:
Initial compilation fix to add a new WatchOS framework target

ertang2015 committed:
Expose an API to acquire token with existing cached auth info

ertang2015 committed:
Replacing the NSURLConnection in ADWebRequest with NSURLSession and add it to WatchOS target

ertang2015 committed:
Fix the NSURLSession issue in WatchOS caused by NSURLProtocol

ertang2015 committed:
- Adding proper comments;
- Remove ADNTLMHandler & ADURLProtocol from WatchOS target compilation;

ertang2015 committed:
nit: Remove unnecessary spaces.

SergeyADoroshenko commented on ADAL/ADAL.xcodeproj/project.pbxproj:
```diff
@@ -2886,6 +3050,273 @@
 			};
 			name = Release;
 		};
+		B681E9151D380633001157CE /* Development */ = {
```
Should Azure have settings for Development, Beta, AppStore, and Nightly?

ertang2015 replied:
No. But this is a todo repository and it's modified to adapt to the todo project settings. In the case when we need to send the pull request back against main open source repository, we need to get rid of all these changes.

atdrendel replied:
Exactly right.

SergeyADoroshenko commented on ADAL/src/request/ADAuthenticationRequest+WebRequest.m:
```diff
@@ -174,6 +176,12 @@ - (NSString*)generateQueryStringForRequestType:(NSString*)requestType
 {
+#if TARGET_OS_WATCH
+    @throw [NSException exceptionWithName:@"WebViewException"
+                                   reason:@"Webview is not support on apple watch"
```
typo: "not supported"

atdrendel commented on ADAL/src/ADAL_Internal.h:
```diff
@@ -54,10 +54,12 @@
 #   include <UIKit/UIKit.h>
 typedef UIWebView WebViewType;
+#elif TARGET_OS_WATCH
+typedef NSString WebViewType; // Should never be called
```
Should this be `id`? Giving it an actual type doesn't make much sense to me. Importantly, I think the accessor should be overridden to throw an exemption if called from on Apple Watch.

ertang2015 replied:
Changed it to a more generic NSObject. Since id is a generic pointer type, if we use id here, we will have compilation issue when declare WebViewType properties in other source code like "@property (weak) WebViewType\* webView;"

atdrendel replied:
Good point and good fix.

ertang2015 committed:
- Replaced WebViewType from NSString to a more generic NSObject type for watchOS;
- removed some compilation warnings for watchOS target

atdrendel commented on ADAL/src/request/ADWebRequest+NSURLSession.h:
```diff
@@ -0,0 +1,71 @@
+}
+
+@property (strong, readonly, nonatomic)     NSURL               *URL;
+@property (copy, nonatomic)                 NSMutableDictionary *headers;
```
Making this copy will always convert this into an `NSDictionary` when using property syntax, not an `NSMutableDictionary`. It should be `strong`.

ertang2015 replied:
The ADWebRequest+NSURLSession is a "copy" of ADWebRequest source code except the usage of NSURLSession, especially the declaration in the header files are pretty much the same. If this would be a problem, should we surface it back to main repository?

atdrendel replied:
Yeah, I think we should fix it in ADWebRequest as well. It’s not a huge deal because this library usually only accesses instance variables directly rather than through property accessors. Still, it’s wrong, and I think we should fix it.

On Jul 29, 2016, at 7:26 PM, Eric Tang <notifications@github.com<mailto:notifications@github.com>> wrote:

In ADAL/src/request/ADWebRequest+NSURLSession.hhttps://github.com/msftberlin/azure-activedirectory-library-for-objc/pull/2#discussion_r72877551:

> -    NSHTTPURLResponse \* _response;
> -    NSMutableData \* _responseData;
>   +
> -    NSUUID \* _correlationId;
>   +
> -    NSUInteger _timeout;
>   +
> -    NSOperationQueue \* _operationQueue;
>   +
> -    BOOL _isGetRequest;
>   +
> -    void (^_completionHandler)(NSError *, ADWebResponse *);
>   +}
>   +
>   +@property (strong, readonly, nonatomic)     NSURL               *URL;
>   +@property (copy, nonatomic)                 NSMutableDictionary *headers;

The ADWebRequest+NSURLSession is a "copy" of ADWebRequest source code except the usage of NSURLSession, especially the declaration in the header files are pretty much the same. If this would be a problem, should we surface it back to main repository?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHubhttps://github.com/msftberlin/azure-activedirectory-library-for-objc/pull/2/files/f8fff06aeb287b8237660a2228bfaaeb16ab8af0#r72877551, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AAMWooo_HpttQOGFH7wyGBqUGJIeVcUAks5qapodgaJpZM4JP-vI.

ertang2015 replied:
Did some experiment and looks like accessing a "copy" property will not create a immutable copy but set will do. 
I think your argument is correct here and I've updated both of the property definition to "strong".

ertang2015 committed:
Merge branch 'master' into ertang/ADALForWatch

ertang2015 committed:
Add ADKeyChainUtil into WatchOS framework compilation sources

atdrendel commented on ADAL/src/request/ADWebRequest+NSURLSession.m:
```diff
@@ -0,0 +1,218 @@
+    }
+    
+    NSURL* requestURL = [ADHelpers addClientVersionToURL:_requestURL];
+    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:requestURL
```
It looks like we'll be leaking `request` when not using ARC.

ertang2015 replied:
Thanks. Will fix it.

ertang2015 replied:
added proper "release" to request object

atdrendel commented:
I added a few comments about small issues. After addressing those, I think this should be merged.

ertang2015 committed:
- Safely release the local request object inside the ADWebRequest+NSURLSession for watchOS;
- Update the "headers" property of ADWebRequest (for both iOS & WatchOS) to strong;

ertang2015 commented:
@atdrendel I fixed the issues you mentioned. Could you take a look at it to see if it's ok to merge?

atdrendel commented:
It's already on my todo list for today. ?

[Image]

atdrendel commented on ADAL/src/request/ADWebRequest.h:
```diff
@@ -47,7 +47,7 @@
 
 @property (strong, readonly, nonatomic) NSURL               *URL;
-@property (copy, nonatomic)             NSMutableDictionary *headers;
+@property (strong, nonatomic)           NSMutableDictionary *headers;
```
👍

atdrendel commented:
From my perspective, it's good to go. Please merge.

atdrendel commented:
I'm going to merge this because I'm block by it.
